### ISSUE TITLE
[Feature Request] config.json: handle null and "" separately

### ISSUE THREAD

**Issue by NP-Hardass** [keeweb/keeweb#669]
I'm running 1.5.4.  I think it would be nice, for certain fields, to handle a value of null and a value of "" differently.  For example, "" would explicitly mean an empty string, and null would mean either use default value, or prompt user for value.

A possible benefit to handling it this way would be that you could serve up a config.json where a webdav uri is given, but user and password are kept as null, so the user would be prompted to manually fill those in when importing the config file.

#668 is a related, but separate feature request.

**Comment by antelle** [comment]
No, it's not obvious. It's better to add a separate property, like askPassword, or whatever. What are you trying to achieve, which peoperty is missing for you now?

**Comment by NP-Hardass** [comment]
@antelle What I am basically trying to achieve is the ability to serve up a config file that doesn't have the password (or username) hardcoded.  The idea is basically that you serve up the webdav uri but don't necessarily force a username or password and let the user fill those in themselves.

Not a problem if this proposal isn't the best way to approach it.  Was just attempting to brainstorm.

**Comment by antelle** [comment]
Yeah this makes sense, we should add an option to ask for password

**Comment by krisjans** [comment]
Also ask user for webDAV password if server reply with error (like HTTP 401) when opening the app;

**Comment by JodliDev** [comment]
Since this issue has been around for some time now: Are there any plans to implement this?

I was looking into implementing something like this for you, but it seems to be more diffictult than I hoped.
Any pointers where to start? I know there is storage-webdav.js but as far as I understand, this is only used when the request is already about to happen (and no more ui input is required).

The config seems to be loaded in applyUserConfig() in app-model.js. 
An alternative idea I had, was maybe have another Array in the config that is similar to files but instead would have the webdav dialog opened and prefill the given values. But I am unsure where to start such an endevor (or if it is even feasible).

**Comment by antelle** [comment]
Eventually it will be implemented, but it's indeed not easy. I'd suggest not doing it now because it would require a bit of refactoring.

**Comment by JodliDev** [comment]
Thank you for you fast response.
Then I will try to be patient :-)
Do you have any suggestion on how to "bookmark" my urls (and possibly the username) on the server without having to supply my webDAV password (since I have other stuff on my webDAV server as well)?

**Comment by antelle** [comment]
Maybe creating a separate WebDAV credentials for KeeWeb can help in this case?

**Comment by JodliDev** [comment]
But that would still expose my database file itself also to people without my webdav password. I dont want to loose that additional layer of protection.

I guess I will create the config.json dynamically with the password field depending on other external input.

Thank you for your help and I am looking forward to this being implemented :)

**Comment by JodliDev** [comment]
In case it is helpful for other people:
If you want to automatically load different keeweb configs that **only** are loaded when a correct password is provided (in my case: I want to load a webDAV file but want to make sure that neither the webDAV password nor the kdbx file can be downloaded without the correct password), I wrote a little hack that might be usefull:

My reasons to go this route:
- The keeweb source does not need to be changed at all.
- There is no need for a validation on the server side (if keeweb works, then this hack works too).
- There can be different settings (e.g. webdav urls) with different passwords.
- Neither the kdbx file nor the webDAV password are exposed to the web until its correct password is entered.
- At no point is the password visible to the user (unless developer console is used, but then it would be visible anyway).

How does it work:
- On an additional login page, it asks the user for a password.
- From this password it derives the filename for a config json (using base64 to make sure special chars are not messing with the system).
- It then opens keeweb, and instructs it to load said config json.
	- The config json is loaded through the meta-tag and not the url to make sure that the password will not be saved to browser history and it is not visible to the user.
- If the config json exists on the server, then the password was correct and the webDAV password is loaded.
- If the password was wrong, keeweb will give out an error and the config json with the webDAV password stays hidden.

How to implement (Assuming, that you want keewep running at `https://example.com`):
- Make sure that keeweb is in a **subfolder** called `keeweb` (e.g. `https://example.com/keeweb`).
- Choose a password, turn the password into a base64 String, then create a config file with all the secret settings you want (see <https://github.com/keeweb/keeweb/wiki/Configuration#json-app-config>) and name the file `BASE64_PASSWORD.json` (e.g. "test" becomes `https://example.com/dGVzdA==.json`).

Then create a file named `index.html` (or whatever you want) with the following content:

```
<!DOCTYPE html><html>
<head>
<meta charset="UTF-8">
<base href="keeweb/">
<script type="text/javascript">
let form, keeWeb_source;
function init() {
	form = document.getElementById("form");
	form.addEventListener("submit", login);
	window.addEventListener("hashchange", onhashchange);
}

//preparing to load config constructed out of password and load keeweb:
function login(e) {
	let pwEl = document.getElementById("password");
	let pw = btoa(pwEl.value); //make password not (quickly) readable for humans
	pwEl.value = "";
	console.log("We will use the config: "+pw+".json");
	document.getElementById("config").setAttribute("content", "../"+pw+".json"); //set config for keeweb
	document.getElementById("submit").disabled = true; //give at least some feedback that we are loading
	
	//load keeWeb:
	let r = new XMLHttpRequest();
	r.open("get", "index.html", true); //because of the base-tag we are actually loading keeweb/index.html
	r.onreadystatechange = function() {
		if(r.readyState !== 4)
			return;
		
		document.getElementById("submit").disabled = false; //enable form again
		if(r.status !== 200)
			alert(r.statusText);
		else {
			keeWeb_source = r.responseText;
			document.location.hash = "#keeweb"; //enable backbutton; keeweb will be inserted in onhashchange()
		}
	}
	r.send(null);
	e.preventDefault(); //dont load a new page after form submit
}

//lets make use of the back & forward buttons:
function onhashchange() {
	switch(document.location.hash) {
		case "#keeweb":
			if(form.parentNode) {
				form.parentNode.removeChild(form); //remove form
				let keeWeb = document.createRange().createContextualFragment(keeWeb_source);
				document.body.appendChild(keeWeb); //insert keeweb
			}
		break;
		default:
			if(!form.parentNode) {
				document.getElementById("config").setAttribute("content", "(no-config)"); //empty password
				document.body.innerHTML = ""; //remove keeweb
				document.body.appendChild(form); //insert form
			}
		break;
	}
}
</script>
<meta id="config" name="kw-config" content="(no-config)">
</head>
<body onload="init()">
<form id="form">
	<input type="password" id="password" placeholder="password"/>
	<input type="submit" value="load config" id="submit"/>
</div>
</body>
</html>
```
- Open `https://example.com`, provide the correct password - Voilá!
